serverjs/cubefn.js

```diff
@@ -228,6 +231,218 @@ async function getElo(cardnames, round) {
+      tags: split[10] && split[10].length > 0 ? split[10].split(',') : [],
+    };
+
+    let potentialIds = carddb.allIds(card);
+    if (potentialIds && potentialIds.length > 0) {
+      // First, try to find the correct set.
+      let matchingSetAndNumber = potentialIds.find((id) => {
+        const dbCard = carddb.cardFromId(id);
+        return (
+          card.set.toUpperCase() == dbCard.set.toUpperCase() &&
+          card.collector_number.toUpperCase() == dbCard.collector_number.toUpperCase()
+        );
+      });
+      let matchingSet = potentialIds.find((id) => carddb.cardFromId(id).set.toUpperCase() == card.set);
+      let nonPromo = potentialIds.find(carddb.notPromoOrDigitalId);
+      let first = potentialIds[0];
+      card.cardID = matchingSetAndNumber || matchingSet || nonPromo || first;
+      if (maybeboard == 'true') {
+        newMaybe.push(card);
+      } else {
+        newCards.push(card);
+      }
+    } else {
+      missing += card.name + '\n';
+    }
+  }
+  return { newCards: newCards, newMaybe: newMaybe, missing: missing };
+}
+
+async function generateBlogpost(req, res, cube, changelog, added, missing, carddb) {
+  var blogpost = new Blog();
+  blogpost.title = 'Cube Bulk Import - Automatic Post';
+  blogpost.html = changelog;
+  blogpost.owner = cube.owner;
+  blogpost.date = Date.now();
+  blogpost.cube = cube._id;
+  blogpost.dev = 'false';
+  blogpost.date_formatted = blogpost.date.toLocaleString('en-US');
+  blogpost.username = cube.owner_name;
+  blogpost.cubename = cube.name;
+
+  if (missing.length > 0) {
+    const reactProps = {
+      cubeID: req.params.id,
+      missing,
+      added: added.map(({ _id, name, image_normal, image_flip }) => ({ _id, name, image_normal, image_flip })),
+      blogpost: blogpost.toObject(),
+    };
+    res.render('cube/bulk_upload', {
+      reactHTML:
+        NODE_ENV === 'production'
+          ? await ReactDOMServer.renderToString(React.createElement(BulkUploadPage, reactProps))
+          : undefined,
+      reactProps: serialize(reactProps),
+      cube: cube,
+      cube_id: req.params.id,
+      title: `${abbreviate(cube.name)} - Bulk Upload`,
+    });
+  } else {
+    blogpost.save(function(err) {
+      cube = setCubeType(cube, carddb);
+      Cube.updateOne(
+        {
+          _id: cube._id,
+        },
+        cube,
+        function(err) {
+          if (err) {
+            req.flash('danger', 'Error adding cards. Please try again.');
+            res.redirect('/cube/list/' + req.params.id);
+          } else {
+            req.flash('success', 'All cards successfully added.');
+            res.redirect('/cube/list/' + req.params.id);
+          }
+        },
+      );
+    });
+  }
+}
+
+// prices should be the prices module with the GetPrices function.
+// elo should be in the form { round: bool }.
+// requested details is a string to pass to carddb.cardFromId.
+async function populateCardDetails(
```

Comment:
I think this function is a lot more complicated than it needs to be. I don't think it needs to do the allDetails thing in addition to everything else, since that's only used in one place. I'm worried about it being written in such an imperative style - instead, I'd rather have it as a pure function. I can give you some suggestions on the discord

Reply:
I think this PR is starting to suffer from trying to do too much - refactor routes, replacing duplicate code with helper functions, in addition to the new csv import. Going forward let's try to create more granular PRs - this will also make code reviews a lot easier, and conflicts less likely